### ISSUE TITLE
salt fileserver does not like cache with date in future

### ISSUE THREAD

**Issue by sjorge** [saltstack/salt#48360]
### Description of Issue/Question
I dropped a new beacon I am working on in file_root/_beacons (also tried _beacon).
```salt \* saltutil.sync_beacons``` or ```salt \* saltutil.sync_all``` are not picking it up.

Checking the local cache on the minion where they should en up, I do not see a _beacons directory. Other directories for _modules and _states show up and get synced correctly.

### Setup
n/a

### Steps to Reproduce Issue
Create a _beacons folder in your file_root, add a new beacon to it. Try to sync it to all minions.

### Versions Report
```Salt Version:
           Salt: 2018.3.1

Dependency Versions:
           cffi: 1.11.5
       cherrypy: 14.0.1
       dateutil: 2.7.2
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 2.1.9
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
        libnacl: 1.6.1
       M2Crypto: 0.29.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.14 (default, Apr  7 2018, 14:51:22)
   python-gnupg: 0.4.2
         PyYAML: 3.12
          PyZMQ: 17.0.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: 0.2.4
        Tornado: 4.5.3
            ZMQ: 4.2.3

System Versions:
           dist:
         locale: UTF-8
        machine: i86pc
        release: 5.11
         system: SunOS
        version: Not Installed```

**Comment by gtmanfred** [comment]
I am unable to replicate this.


```
[root@salt ~]# cat /srv/salt/_beacons/this.py
'''
TEST
'''
[root@salt ~]# salt \* saltutil.sync_beacons
salt:
    - beacons.this
[root@salt ~]# cat /var/cache/salt/minion/extmods/beacons/this.py
'''
TEST
'''
```

Are you putting the beacon in the 'base' fileroot? or another file root?

**Comment by sjorge** [comment]
I have 2 file roots, one just has all the extra modules.

Everything is picked up except the beacons.

**Comment by sjorge** [comment]
There is no difference if I put them in /salt/states/_beacons, that results in the same thing.

**Comment by gtmanfred** [comment]
Unfortunately I am unable to replicate this.

Can you do `salt-call cp.list_master prefix=_beacons` and see what it returns?

Also, check the minion extmod cache directory.

Thanks!
Daniel

**Comment by sjorge** [comment]
```
[cronos :: sjorge][~]
[.]$ salt-call cp.list_master prefix=_beacons
local:

## weird, checking _modules to for comparison
[cronos :: sjorge][~]
[.]$ salt-call cp.list_master prefix=_modules
local:
    - _modules/README
    - _modules/smartos_imgadm.py

## checking find
[cronos :: sjorge][~]
[.]$ find /salt/states/.extmods 
/salt/states/.extmods
/salt/states/.extmods/_states
/salt/states/.extmods/_states/smartos.py
/salt/states/.extmods/_states/README
/salt/states/.extmods/_modules
/salt/states/.extmods/_modules/README
/salt/states/.extmods/_modules/smartos_imgadm.py
/salt/states/.extmods/_modules/smartos_vmadm.py
/salt/states/.extmods/_runners
/salt/states/.extmods/_runners/README
/salt/states/.extmods/_beacons
/salt/states/.extmods/_beacons/smartos_imgadm.py
/salt/states/.extmods/_grains
/salt/states/.extmods/_grains/README
/salt/states/.extmods/_grains/mdata.py
/salt/states/.extmods/_utils
/salt/states/.extmods/_utils/README
```

So salt is not even finding the directory!

**Comment by sjorge** [comment]
Interesting! I noticed all my other modules are now also no longer synced! Event those that show up with cp.list_master!

**Comment by sjorge** [comment]
Ok so cp.list_master is not picking up any new files I add... anyway to debug this to figure out why? Nothing in the log for master/minion.

**Comment by gtmanfred** [comment]
Well, that explains why it isn't syncing them... now to figure out why they aren't finding them...

**Comment by sjorge** [comment]
I removed file_list and roots from /var/cache/salt/master and now cp.list_master picks up the _beacons directory.

```
# salt carbon saltutil.sync_beacons
carbon:
    - beacons.smartos_imgadm_beacon
```

off-topic: The amount of type I have typed `bacon` instead of beacon is ridiculous!

**Comment by gtmanfred** [comment]
That is super bizarre... so it is working now?

**Comment by gtmanfred** [comment]
Also yes, i have typed bacon a couple times too. ha

**Comment by sjorge** [comment]
It's working now, and picking up new test files as I add them. Based on the timestamps of those dirs I cleared in the cache. They have not been update since I switched to 2018.3.1 from 2018.3.0...

I think we can close this for now, but still super weird it stopped picking up new files. Could one of the cache files in there have gotten corrupt?

**Comment by gtmanfred** [comment]
That I do not know, @terminalmage can you think of something that might have happened with the fileserver here?

Thanks,
Daniel.

**Comment by sjorge** [comment]
Ok this sucks, salt stopped picking up new files again. After clearing those 2 dirs again it started working again?!

**Comment by sjorge** [comment]
Giving everything a `find /var/cache/salt/master -exec touch {} \;` also seems to fix the issue!

So something is probably doing a timestamp check? I had a quick look in the code but could not immediatly see where.

**Comment by gtmanfred** [comment]
My understanding is they should be synced if the hash changes, and touching a file should change the mtime, and I believe, change the hash.

Not sure why they would have stopped syncing before that though.

**Comment by sjorge** [comment]
In this case the hash timestamp was newer than the file timestamps due to the ntp issues... something that generates/checks the cache/hash is probably also checking timestamps.

**Comment by terminalmage** [comment]
@sjorge if it's a brand new file, then the fileclient would have no reason to assume that the file was already downloaded, and would not skip caching it.

My guess is that you're running into an issue with the file list cache. Salt caches the results of getting the list of available files for a given environment, because a thousand minions all asking the master to walk `/srv/salt` at the same time is not nice to the master's CPU. This cache ages out by default after 20 seconds. This can be tuned by using the [fileserver_list_cache_time](https://docs.saltstack.com/en/latest/ref/configuration/master.html#fileserver-list-cache-time) config option, and those docs I linked also point to a runner function which can be used to clear these caches.

I think the issue with the timestamps was probably a red herring, and that by the time you tried something else, the file list cache had already aged out.

**Comment by sjorge** [comment]
The original one went days without updating it, how does it check the age of the cache file? if the timestamp is due to a ntp bug say 2039... does it age out before 2039? From what I see it does not.

(We should probably update the title of this bug though)

**Comment by sjorge** [comment]
Sweet I got it down to a minimal reproducable case

```
# mark the cache as modified in the far future
[cronos :: sjorge][/var/cache/salt/master]
[.]$ pfexec find file_lists/ -exec touch -m -t 201912171200 {} \;
[cronos :: sjorge][/var/cache/salt/master]
[.]$ pfexec find roots -exec touch -m -t 201912171200 {} \;
# drop in a new module
[cronos :: sjorge][/var/cache/salt/master]
[.]$ echo "'''Hello'''" > /salt/states/.extmods/_modules/hello.py
# sync finds nothing
[cronos :: sjorge][/var/cache/salt/master]
[.]$ salt cronos saltutil.sync_all
cronos:
    ----------
    beacons:
    clouds:
    engines:
    grains:
    log_handlers:
    modules:
    output:
    proxymodules:
    renderers:
    returners:
    sdb:
    states:
    thorium:
    utils:
```
It does appear to be the case for all files that go through the file server.

**Comment by sjorge** [comment]
I will create a PR later this weekend with a fix or at least with a warning

**Comment by gtmanfred** [comment]
perfect, thanks @sjorge

**Comment by rallytime** [comment]
Closed via #48478
